### PR TITLE
ATL dependency graph

### PR DESCRIPTION
**Pull request by jenrik**
Implements the dependency graph for ATL as currently described in our worksheets.

The ALWAYS expression has been omitted because it can be derived using the other formulas (or so says @NicEastvillage).

### CODE REVIEW

File: src/atl/dependencygraph.rs [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+                                    })
+                                    .collect();
+                            Edges::HYPER(HyperEdge {
+                                source: vert.clone(),
+                                targets,
+                            })
+                        })
+                        .collect::<HashSet<Edges<ATLVertex>>>()
+                }
+                Phi::UNTIL {
```
Comment (NicEastvillage) [inline src/atl/dependencygraph.rs:315]: Så vidt jeg kan se, har du prøvet at lave despite-until (Figur 6), men du har ikke fået inverteret spillerne. Bemærk at det er de *andre* spillere, der bruges i `vars` i sidebetængelsen fra Figur 6.

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
Found some things. Otherwise it looks good!

File: src/atl/formula.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    NEXT {
+        players: Vec<Player>,
+        formula: Arc<Phi>,
+    },
+    UNTIL {
+        players: Vec<Player>,
+        pre: Arc<Phi>,
+        until: Arc<Phi>,
+    },
+}
```
[the remark this answers is not in the document]
Comment (jenrik) [reply src/atl/formula.rs:18]: > The always formula can be derived from DESPITE_UNTIL.

How? Where is that described in the worksheet?

File: src/atl/dependencygraph.rs (hunk 1 above)
Comment (jenrik) [reply src/atl/dependencygraph.rs]: fixed in 0a41d11f149a2df0c0a4a134b87262e940282353

File: src/atl/formula.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jenrik) [reply src/atl/formula.rs]: DESPITE have been renamed to DESPITE_UNTIL in 8e5254fa686e01ab7610b7bdedae25323692f34a

File: src/atl/dependencygraph.rs [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+                    let right_targets = vec![ATLVertex::FULL {
+                        state: *state,
+                        formula: right.clone(),
+                    }];
+                    edges.insert(Edges::HYPER(HyperEdge {
+                        source: vert.clone(),
+                        targets: right_targets,
+                    }));
+
+                    edges
```
[the remark this answers is not in the document]
Comment (jenrik) [reply src/atl/dependencygraph.rs]: fixed in 1105769770641d803da46e040f75b1ef85350450

File: src/atl/formula.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jenrik) [reply src/atl/formula.rs]: ENFORCE_UNTIL have been implemented in 68012e1fbe44dba72676d7e168e448cf5f2e7a92

**Review by NicEastvillage — APPROVED** [review]
Awesome

File: src/atl/gamestructure/mod.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+mod eager;
+
+#[allow(unused_imports)]
+pub(crate) use eager::EagerGameStructure;
+
+pub(crate) trait GameStructure {
+    fn max_player(&self) -> u32;
+
+    fn labels(&self, state: usize) -> HashSet<Proposition>;
```
Comment (NicEastvillage) [inline src/atl/gamestructure/mod.rs:12]:
```suggestion
    fn labels(&self, state: State) -> HashSet<Proposition>;
```

File: src/atl/gamestructure/eager.rs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    fn available_moves(&self, state: usize, player: usize) -> u32 {
+        *self
+            .moves
+            .get(state)
+            .unwrap_or_else(|| panic!("Requested move for non-existent state {}", state))
+            .get(player)
+            .unwrap_or_else(|| panic!("Request move for non-existent player {} from state {}", player, state))
+    }
+
+    fn move_count(&self, state: usize) -> Vec<u32> {
```
Comment (NicEastvillage) [inline src/atl/gamestructure/eager.rs:49]:
```suggestion
    fn move_count(&self, state: State) -> Vec<u32> {
```

File: src/atl/gamestructure/eager.rs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        transition_lookup(
+            choices.as_slice(),
+            &self
+                .transitions
+                .get(state)
+                .unwrap_or_else(|| panic!("Undefined state {}, no transitions", state)),
+        )
+    }
+
+    fn available_moves(&self, state: usize, player: usize) -> u32 {
```
Comment (NicEastvillage) [inline src/atl/gamestructure/eager.rs:40]:
```suggestion
    fn available_moves(&self, state: State, player: Player) -> u32 {
```

File: src/atl/gamestructure/eager.rs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    /// available moves for a player in a given state
+    pub moves: Vec<Vec<u32>>,
+}
+
+impl GameStructure for EagerGameStructure {
+    fn max_player(&self) -> u32 {
+        self.player_count
+    }
+
+    fn labels(&self, state: usize) -> HashSet<Proposition, RandomState> {
```
Comment (NicEastvillage) [inline src/atl/gamestructure/eager.rs:23]:
```suggestion
    fn labels(&self, state: State) -> HashSet<Proposition, RandomState> {
```

**Comment by jenrik** [comment]
[Review ATL dependnecy graph PR](https://trello.com/c/zJmbYs0I/46-review-atl-dependnecy-graph-pr)

**Pull request merged**